### PR TITLE
ensure that clients don't miss ports/tasks status updates

### LINKED ISSUE #2290: tasks/ports updates should not be dropped
Right now some updates can be dropped while the client is subscribing. It is not very problematic for ports, but for tasks it can lead to a situation when a user never see a running terminal.

### PR DESCRIPTION
**Pull request by akosyakov**
#### What it does

- fix #2290: it make sure that no updates can help between clients receiving an initial status and subscribing
- added unit tests for concurrent subscription of ports (since we had issued there in the past)
- fixed task management that run only returns when all tasks are done, before we will miss some conditions like if a task could not be started at all but the prebuild will look successful

#### How to test

- tests still should pass
- terminals and ports should work

### CODE REVIEW

**Review by csweichel — APPROVED** [review]
other than the chance commentary,
this PR looks good to me.

File: components/supervisor/pkg/ports/ports_test.go [hunk 1]
```diff
@@ -366,3 +372,83 @@ type testServedPorts struct {
...
+				return
+			}
+		}
+	}()
+
+	eg, _ := errgroup.WithContext(context.Background())
+	for i := 0; i < maxSubscriptions; i++ {
+		eg.Go(func() error {
+			for j := 0; j < subscribes; j++ {
+				sub, err := pm.Subscribe()
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply components/supervisor/pkg/ports/ports_test.go:430]: yes, it is testing for deadlocks between subscribing and processing events

**Comment by akosyakov** [comment]
I added comments and will merge if the build is green.

**Pull request merged**